azuredevops/time.go

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azuredevops
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+)
+
+type Time struct {
+	Time time.Time
+}
+
+func tryUnmarshalJSON(b []byte) (time.Time, error) {
+	t := time.Time{}
+	err := json.Unmarshal(b, &t)
+
+	// ignore errors for 0001-01-01T00:00:00 dates. The Azure DevOps service
+	// returns default dates in a format that is invalid for a time.Time. The
+	// correct value would have a 'z' at the end to represent utc. We are going
+	// to ignore this error, and set the value to the default time.Time value.
+	// https://github.com/microsoft/azure-devops-go-api/issues/17
+	if err != nil {
+		if parseError, ok := err.(*time.ParseError); ok && parseError.Value == "\"0001-01-01T00:00:00\"" {
+			err = nil
+		}
+	}
+
+	return t, err
+}
+
+func makeTimeUnmarshalStrategyFunc(layout string) func(b []byte) (time.Time, error) {
+	return func(b []byte) (time.Time, error) {
+		s := strings.Trim(string(b), "\"")
+		if s == "null" {
+			return time.Time{}, nil
+		}
+		return time.Parse(layout, s)
+	}
+}
+
+// UnmarshalJSON attempts a variety of time marshalling strategies. The first one that works
+// will be used. If none succeede, then an error is returned.
+func (t *Time) UnmarshalJSON(b []byte) error {
```

Comment:
@nmiodice , I know you found one property that was returning a date in an incorrect format.  Were there others?  The long term fix is to fix the service.  All date formats the service returns should be the same.  If this is the only known case, then we should put in a temporary workaround for that specific case until the issue is fixed.  Thoughts?

Reply:
I do agree that each API should return a consistent time format. In practice, I see no reason to code defensively so that the client is more robust to fixes in the service.

The "fix for each special case" is probably a good way to identify issues in the service, but it comes with a large cost of slowing down development in other projects if manual intervention needs to occur for each new issue found. My vote is to be very lenient with time formats and push on the product team to fix the underlying service bugs.

Reply:
Hey @nmiodice , I really appreciate you taking the time to put this PR together.  I understand your point of view around defensive programming.  For now I would like to keep things as simple as possible and inline with our other SDKs, by just working around this particular azure devops bug.  I think an azure devops developer would really have to go out of there way and create a custom json converter to have the dates from their contracts match most of these formats.  We can revisit if another oddball case comes up.  For now I am going to add a quick workaround, and move on to the other issues.